lxd/acme.go

```diff
@@ -0,0 +1,185 @@
+		}
+
+		// This gives me the correct value
+		clusterAddress := d.State().LocalConfig.ClusterAddress()
+
+		if clusterAddress != "" && clusterAddress != leader {
+			// Forward the request to the leader
+			client, err := cluster.Connect(leader, d.endpoints.NetworkCert(), d.serverCert(), r, true)
+			if err != nil {
+				return response.SmartError(err)
+			}
+
+			return response.ForwardedResponse(client, r)
+		}
+	}
+
+	if d.http01Provider == nil || d.http01Provider.Token() != token {
+		return response.NotFound(nil)
+	}
+
+	return response.ManualResponse(func(w http.ResponseWriter) error {
+		w.Header().Set("Content-Type", "text/plain")
+
+		_, err := w.Write([]byte(d.http01Provider.KeyAuth()))
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+func autoRenewServerCert(ctx context.Context, d *Daemon) error {
+	domain, email, caURL := d.globalConfig.ACME()
+
+	if domain == "" || email == "" {
+		return nil
+	}
+
+	clustered, err := cluster.Enabled(d.db.Node)
+	if err != nil {
+		return err
+	}
+
+	var clusterAddress string
+
+	// If we are clustered, let the leader handle the certificate renewal.
+	if clustered {
+		leader, err := d.gateway.LeaderAddress()
+		if err != nil {
+			return err
+		}
+
+		// Figure out our own cluster address.
+		clusterAddress := d.State().LocalConfig.ClusterAddress()
+
+		if clusterAddress != leader {
+			return nil
+		}
+	}
+
+	opRun := func(op *operations.Operation) error {
+		cert, err := acme.UpdateACMECertificate(d.State(), d.http01Provider, clustered, domain, email, caURL)
+		if err != nil {
+			return err
+		}
+
+		// If cert is nil, there's no need to update it as it's still valid for more than 30 days.
+		if cert == nil {
+			return nil
+		}
+
+		req := api.ClusterCertificatePut{
+			ClusterCertificate:    string(cert.Certificate),
+			ClusterCertificateKey: string(cert.PrivateKey),
+		}
+
+		if clustered {
+			servers, err := d.gateway.NodeStore().Get(context.Background())
+			if err != nil {
+				return err
+			}
+
+			for _, server := range servers {
+				if server.Address == clusterAddress {
+					continue
+				}
+
+				client, err := cluster.Connect(server.Address, d.endpoints.NetworkCert(), d.serverCert(), nil, true)
+				if err != nil {
+					return err
+				}
+
+				err = client.UpdateClusterCertificate(req, "")
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		client, err := cluster.Connect(d.State().LocalConfig.HTTPSAddress(), d.endpoints.NetworkCert(), d.serverCert(), nil, true)
+		if err != nil {
+			return err
+		}
+
+		err = client.UpdateClusterCertificate(req, "")
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	op, err := operations.OperationCreate(d.State(), "", operations.OperationClassTask, operationtype.RenewServerCertificate, nil, nil, opRun, nil, nil, nil)
+	if err != nil {
+		logger.Error("Failed to start renew server certificate operation", logger.Ctx{"err": err})
+		return err
+	}
+
+	logger.Info("Renewing server certificate")
```

Comment:
This is a bit misleading as the certificate isn't necessarily being renewed, how about:

"Automatic server certificate renewal check starting"
"Automatic server certificate renewal check done"